src/components/queryExpressionEditor/advancedMode/AdvancedMode.tsx

```diff
@@ -0,0 +1,50 @@
+import * as React from 'react';
+import {CodeEditor} from '../../editor/CodeEditor';
+import {CodeMirrorModes} from '../../editor/EditorConstants';
+import * as styles from './AdvancedMode.scss';
+
+export interface IAdvancedModeProps {
+    queryExpression: string;
+    updateQueryExpression: (expression: string) => void;
+}
+
+export interface IAdvancedModeState {
+    initialContent: string;
+}
+
+export class AdvancedMode extends React.Component<IAdvancedModeProps, IAdvancedModeState> {
+    private codeEditorContent: string;
+
+    constructor(props: IAdvancedModeProps) {
+        super(props);
+        this.state = {initialContent: ''};
+        this.codeEditorContent = '';

```

Comment:
I agree with what Oli and Sami are saying, codeEditorContent is basically the queryExpression. You should find a way not to duplicate this value in two entities. Why not debouncing input changes and updating the queryExpression after a certain timeout? I think it would work and typing in the code editor would feel smoother.

```
export class AdvancedMode extends React.Component<IAdvancedModeProps, IAdvancedModeState> {
    private codeEditorContent: string;

    constructor(props: IAdvancedModeProps) {
        super(props);
        this.state = {initialContent: ''};
        this.codeEditorContent = '';
    }

    private debouncedUpdateQueryExpression = _.debounce((content: string) => this.props.updateQueryExpression(content), 200);

    render() {
        return (
            <div>
                <div className='text-medium-blue mt4 ml4'>
                    Write your query using <a href='https://onlinehelp.coveo.com/en/cloud/coveo_cloud_query_syntax_reference.htm'>Coveo query syntax </a>.
                </div>
                <div className={`mt3 ml4 ${styles.codeEditor}`}>
                    <CodeEditor
                        value={this.props.queryExpression} mode={CodeMirrorModes.Python}
                        onChange={(content: string) => this.debouncedUpdateQueryExpression(content)}
                    />
                </div>
            </div>
        );
    }
}  

```

Reply:
Also, if there is an existing queryExpression , we should be able to see it on mount/first, currently it seems like it is always an empty.

Reply:
https://underscorejs.org/#debounce